examples/server/tests/features/steps/steps.py

```diff
@@ -0,0 +1,283 @@
+
+
+@step(u'a user prompt {user_prompt}')
+def step_user_prompt(context, user_prompt):
+    context.user_prompt = user_prompt
+
+
+@step(u'a system prompt {system_prompt}')
+def step_system_prompt(context, system_prompt):
+    context.system_prompt = system_prompt
+
+
+@step(u'a model {model}')
+def step_model(context, model):
+    context.model = model
+
+
+@step(u'{max_tokens} max tokens to predict')
+def step_max_tokens(context, max_tokens):
+    context.n_predict = int(max_tokens)
+
+
+@step(u'streaming is {enable_streaming}')
+def step_streaming(context, enable_streaming):
+    context.enable_streaming = enable_streaming == 'enabled' or bool(enable_streaming)
+
+
+@step(u'an OAI compatible chat completions request')
+def step_oai_chat_completions(context):
+    oai_chat_completions(context, context.user_prompt)
+
+
+@step(u'a prompt')
+def step_a_prompt(context):
+    context.prompts.append(context.text)
+
+
+@step(u'concurrent completion requests')
+def step_concurrent_completion_requests(context):
+    concurrent_requests(context, request_completion)
+
+
+@step(u'concurrent OAI completions requests')
+def step_oai_chat_completions(context):
+    concurrent_requests(context, oai_chat_completions)
+
+
+@step(u'all prompts are predicted')
+def step_all_prompts_are_predicted(context):
+    for completion_thread in context.completion_threads:
+        completion_thread.join()
+    for completion in context.completions:
+        assert_n_tokens_predicted(completion)
+
+
+@step(u'embeddings are computed for')
+def step_compute_embedding(context):
+    response = requests.post(f'{context.base_url}/embedding', json={
+        "content": context.text,
+    })
+    assert response.status_code == 200
+    context.embeddings = response.json()['embedding']
+
+
+@step(u'embeddings are generated')
+def step_compute_embeddings(context):
+    assert len(context.embeddings) > 0
+
+
+@step(u'an OAI compatible embeddings computation request for')
+def step_oai_compute_embedding(context):
+    openai.api_base = f'{context.base_url}/v1'
+    embeddings = openai.Embedding.create(
+        model=context.model,
+        input=context.text,
+    )
+    context.embeddings = embeddings
+
+
+@step(u'tokenizing')
+def step_tokenize(context):
+    context.tokenized_text = context.text
+    response = requests.post(f'{context.base_url}/tokenize', json={
+        "content":context.tokenized_text,
+    })
+    assert response.status_code == 200
+    context.tokens = response.json()['tokens']
+
+
+@step(u'tokens can be detokenize')
+def step_detokenize(context):
+    assert len(context.tokens) > 0
+    response = requests.post(f'{context.base_url}/detokenize', json={
+        "tokens": context.tokens,
+    })
+    assert response.status_code == 200
+    print(response.json())
+    # FIXME the detokenize answer contains a space prefix ?
+    assert context.tokenized_text == response.json()['content'].strip()
```

Comment:
@BruceMacD @ggerganov Is it normal that the `/detokenize` enpoint prefix the content with a space ? see ##3287

Reply:
Yes, it is expected - SPM tokenizer adds a whitespace prefix: https://github.com/google/sentencepiece/issues/15